storagenode/pieces/readwrite.go

```diff
@@ -5,25 +5,57 @@ package pieces
 
 import (
 	"context"
+	"encoding/binary"
 	"hash"
 	"io"
 
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/pkcrypto"
 	"storj.io/storj/storage"
 )
 
+const (
+	// V1PieceHeaderSize is the size of the piece header used by piece storage format 1 (.sj1).
+	// It includes the size of the framing field (v1PieceHeaderFrameSize). It has a constant
+	// size because:
+	//
+	//  * we do not anticipate needing more than this
+	//  * we will be able to sum up all space used by a satellite (or all satellites) without
+	//    opening and reading from each piece file
+	//  * this simplifies piece file writing (no need to precalculate the necessary header
+	//    size before writing)
+	//
+	// If more space than this is needed, we will need to use a new storage format version.
+	V1PieceHeaderSize = 512
+
+	// v1PieceHeaderFramingSize is the size of the field used at the beginning of piece
+	// files to indicate the size of the piece header (because protobufs are not self-
+	// delimiting, which is lame).
+	v1PieceHeaderFramingSize = 2
+)
+
 // Writer implements a piece writer that writes content to blob store and calculates a hash.
 type Writer struct {
-	hash hash.Hash
-	blob storage.BlobWriter
-	size int64
+	hash     hash.Hash
+	blob     storage.BlobWriter
+	dataSize int64 // piece size only; i.e., not including piece header
 
 	closed bool
 }
 
 // NewWriter creates a new writer for storage.BlobWriter.
-func NewWriter(blob storage.BlobWriter) (*Writer, error) {
+func NewWriter(blob storage.BlobWriter, formatVersion storage.FormatVersion) (*Writer, error) {
 	w := &Writer{}
+	if formatVersion < storage.FormatV1 {
+		return nil, Error.New("writing to storage format version 0 is not supported")
+	}
+	// skip header area for now; fill in on commit
```

Comment:
any chance we could add to this comment and say why we are skipping the header until later?

Reply:
There is a pretty good chance. Hard to estimate the chance to any good degree of precision.

(Yes I'll totally add that!)

Reply:
lol!